### PR TITLE
device: make sure to always retroactively start device dependencies

### PR DESCRIPTION
**Pull request by fbuihuu**
PID1 updates the state of device units upon 2 different events:

  - when it processes an event sent by udev and in this case the device deps are
    started if the device enters in the "plugged" state.

  - when it enumerates all devices during its startup or when it is asked to
    reload its configuration data but in this case the device deps (if any) are
    not retroactively started.

Whe udev processes a new "add" kernel event, it first registers the new device
in its database then sends a new event to systemd.

If for any reason, systemd is asked to reload its configuration (via
"daemon-reload" command for example) between the previous 2 steps, it might see
for the first time the new device when enumerating all devices via udev DB
first and process the event sent by udev in a second step.

In this peculiar case the device deps wont be started (even though the device
is first seen by PID1).

Indeed while "daemon-reload", PID1 will put the device unit in the "plugged"
state but without starting the device deps. Thereafter PID1 will get the event
from udev for the new device but the device unit will be in "plugged" state
already therefore it won't see any need to start the device dependencies.

Rather than assuming that during the reloading of systemd manager configuration
all devices listed in udev DB have been already processed and should be put in
the "plugged" state (done by device_coldplug()), this patch does that only for
devices which have been processed via an udev event (device_dispatch_io())
previously. In this case we set "d->found" to "DEVICE_FOUND_UDEV" and we make
also sure to no more initialize "d->found" while enumerating devices. Instead
this field is now saved/restored while devices are serialized.

### CODE REVIEW

**Comment by fbuihuu** [comment]
Please double check as I'm not really sure about this one.

File: src/core/device.c [hunk 1]
```diff
@@ -165,6 +165,7 @@ static int device_serialize(Unit *u, FILE *f, FDSet *fds) {
         assert(fds);
 
         unit_serialize_item(u, f, "state", device_state_to_string(d->state));
+        unit_serialize_item_format(u, f, "found", "%i", d->found);
```
Comment (poettering) [inline src/core/device.c]: i feel a bit nervous about serializing the binary value of this flag field, as it turns something into an API that is not necessarily obvious as one... Hence, I'd rather see three individual bool fields serialized for the three relevant bits in the flags field...

Comment (fbuihuu) [reply src/core/device.c]: good point, I'll update the PR accordingly, thanks for the review.

**Comment by fbuihuu** [comment]
New version forced pushed:

- the serialization of "found" does not use the raw bit format anymore
- I added an additional fix in a separate commit (see commit message for details)

File: src/core/device.c [hunk 2]
```diff
@@ -179,6 +179,16 @@ static int device_deserialize_item(Unit *u, const char *key, const char *value,
         assert(value);
         assert(fds);
 
+        /* The device was known at the time units were serialized but it's not
+         * anymore at the time units are deserialized. This happens when PID1 is
+         * reexec after having switched to the new rootfs. When devices are
```
Comment (keszybz) [inline src/core/device.c]: reexec → reexecuted

File: src/core/device.c [hunk 3]
```diff
@@ -179,6 +179,16 @@ static int device_deserialize_item(Unit *u, const char *key, const char *value,
...
+         * enumerated, udevd isn't running yet and therefore the list of devices
+         * (handled by systemd) to init was empty. We ignore completely such
+         * case and wait for the device events to be re-triggered by udev so
```
Comment (keszybz) [inline src/core/device.c]: Hmm... "we completely ignore such case" — I guess you mean the opposite, we handle this case (by doing nothing now and waiting for the device to be re-triggered by udev and initialized properly).

Comment (fbuihuu) [reply src/core/device.c]: yes that's what I tried to mean, will reword this part. Thanks.

**Comment by fbuihuu** [comment]
@keszybz you're right it's currently broken when PID is reexecuted.

That was actually the point I raised in my previous comment https://github.com/systemd/systemd/pull/8675#issuecomment-381663923. And I wondered how it was supposed to be handled since all new introductions of field serialization  have the same issue...

**Comment by keszybz** [comment]
> That was actually the point I raised in my previous comment #8675 (comment)

I was responding to that ;)

> And I wondered how it was supposed to be handled since all new introductions of field serialization have the same issue...

Right. We always need to do something backwards compatible, as far as reasonable. Sometimes it's not possible, vide e8a565cb660a7a11f76180fe441ba8e4f9383771.

**Comment by fbuihuu** [comment]
@keszybz how about something like this ?

**Comment by keszybz** [comment]
`KVM: entry failed, hardware error 0x0`

**Comment by keszybz** [comment]
Looks good.

**Pull request merged**